### ISSUE TITLE
Make two more fields multiselect

### ISSUE THREAD

**Issue by Woseseltops** [Signbank/Global-signbank#762]
Currently you can select more than 1 value for the field *Regional Dialect*, Julie and @ocrasborn are thinking about also wanting this for the fields:
* Semantic field
* Derivation history (ASL only)

**Comment by susanodd** [comment]
Currently the Semantic Field is a choice field, and the Dialect field is a ManyToMany field with the model Dialect.
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/dictionary/models.py#L662
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/dictionary/models.py#L493
So this issue entails creating a new model for Semantic Fields.

That's interesting. I just did a search on other ManyToMany fields.
The Gloss field signlanguage is also ManyToMany.
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/dictionary/models.py#L473
But we don't seem to exploit that because a gloss only belongs to one dataset (via its lemma).
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/dictionary/models.py#L470
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/dictionary/models.py#L2337-L2338

**Comment by susanodd** [comment]
I'm working on this issue now.
In order to implement multiple select, as discussed above, a new model SemanticField is needed.
Another field that is multiple select is Dialect.

At the moment, the semField of Gloss is a field choice category.
This choice list can be seen in Admin. It has the FieldChoice fields: machine_value, and the hard coded language fields for Dutch, English, and Chinese.
Recall that Handshapes also were once FieldChoice objects. When the Handshape model was created, new fields were added specific to this new model. The original FieldChoice fields machine_value and the hard coded Dutch, English and Chinese language fields were maintained to support compatibility.

@ocrasborn what extra fields would be useful in a SemanticField class?
In what sense are they multilingual?
What language(s) should be included in a semantic field object?
Would it be at all relevant to let Django do the translation part of this?
There are many more languages now than in the beginning.

I'll start out just putting the current choices for semField into a new class SemanticField.

However, I'm wondering what will happen if people start creating more semantic fields in specific languages rather than relying on translations of the ones that already exist? It could end up that inter-language comparisons are no longer possible as one would need to specifically select all the translations for a desired semantic field, rather than just the semantic field of interest.

**Comment by susanodd** [comment]
Any objection to a menu item "Show Semantic Fields" under say the Analysis menu?

The Handshapes are available via a Handshape Search page, which starts out as listing all the handshapes.
The HandshapeDetailView is actually where the Handshape objects are first created from the Handshape FieldChoice objects.
So if we originally have a Semantic Field List View, then they can be displayed. Then if you select one to go to Semantic Field Detail View, then the new object can be created and you would be able to add desired translations.

**Comment by susanodd** [comment]
Okay, I created two model classes, SemanticField and SemanticFieldTranslation.

There's a SemanticFieldDetailView that takes the machine_value of the original semField FieldChoice in the url.
The data stored in FieldChoice is used to create a new SemanticField object. This is done behind the scenes.

This is what the interface.looks like so far. At the moment, there are the translations for Dutch and Chinese (from FieldChoice).
This will be modified so users can add translations for other languages for the semantic field.

![semanticfield_detailview_37](https://user-images.githubusercontent.com/24272544/150554337-78d12ea5-a13b-488e-b5a9-f459a223b70b.png)

**Comment by susanodd** [comment]
Okay, I've added a list view with links to the detail views.

![semanticfields_list](https://user-images.githubusercontent.com/24272544/150783708-3c30b5d3-e6f9-4c69-9814-b0d67d27716f.png)

![semanticfield-detail-clothing](https://user-images.githubusercontent.com/24272544/150783729-bd398144-0ea6-435f-b3a1-021a2bef4f11.png)

**Comment by susanodd** [comment]
Oh, I should clarify what the above has to do with the "multiselect" of this issue.
The above screenshots show the added classes needed to implement multiple select, as described above.
I've made a new field in Gloss, 'semFieldShadow'. This will exist in parallel to the current 'semField' field.
The new shadow field supports multiple select.
The things which can be multiple selected are shown in the screenshots above.
These exist in parallel with the resp. FieldChoices of type semField.

**Comment by susanodd** [comment]
I've implemented the Admin for the two new classes.

**Comment by susanodd** [comment]
I pushed a working version of the code.

I added Admin functionality. It's now possible to create Semantic Fields in the Admin, along with translations.

I left the origin Field Choice semField as is. There is a new gloss field that is multiselect. The original field is hidden in the Gloss Detail View and only the new multiselect version is shown.

I left the choice lists in the Search panels to keep using the semField "field choices".

As is done for Handshapes, the new SemanticField model lives along side the FieldChoice SemField choices.
There is redundancy conversion built in. Thus the original SemField choices are mapped (created) to SemanticField objects and their respective translations. Also going the other way. So if a user creates a new FieldChoice for SemField in the Admin, or if the user creates a new SemanticField in the Admin, these both get mapped to the other representation.

The mapping (conversion/creation) between the two is done via the new SemanticField Detail View and List View.

I will work on adding more functionality to these.

It works now to add translations for any language for a particular Semantic Field. So far, this is available in the Admin.

**Comment by susanodd** [comment]
This is more or less ready to go. (See screenshots above.)

The main improvement that could be done, would be to allow editing of the Semantic Field objects inside of the detail view.
Namely, to allow adding translations. This can be done in Admin, of course.

The idea is to put the Semantic Field List underneath Analysis.
It could also be extended to show glosses with the particular semantic field in the detail view.
This could be done cross-lanuage (some datasets have more semantic fields specified).

@ocrasborn any ideas?
Is this sufficient to support what will be done with the semantic fields?

So far, only the Semantic Field has become multiselect.
The Derivation History still needs to be done.

**Comment by susanodd** [comment]
I modified the get_queryset functionality of GlossList and MorphemeList to access the new multiselect semantic field.

**Comment by susanodd** [comment]
I'm implementing some tests for the new multi-select semantic field.

**Comment by susanodd** [comment]
There is a pull request on the `semantic_fields` branch #794.

I'll implement the other request for multiselect Derivation History, although the issue says ASL only.

**Comment by susanodd** [comment]
@Woseseltops I'm making a lot of progress on the derivation history. It's analogous to the Semantic Fields implementation.
Although since they're currently not used in Global, I've made a url `configure_derivationhistory` to fill the field choices with the  derivHist choices from the original migration creation, in order to not leave them empty. These can then be edited in Admin.

**Comment by susanodd** [comment]
Here's what it looks like for derivation histories. This is implemented analogously to semantic fields.
(Edit views for Gloss and Morpheme, Search Multiselect, and Derivation History List View and Detail View.)
If there are no derivation history field choices yet, these are created based on the original Global values of the initial migration.
The Admin can then modify these as needed.

I've added a setting `USE_DERIVATIONHISTORY` and set it to True on my local server. The idea is that should this not be wanted (False) that these will be hidden in the interface.


![multisel_derivHist](https://user-images.githubusercontent.com/24272544/160627852-f4068b50-3cbe-4f58-8be5-cae62ca36551.png)

![multisel_derivHist_search](https://user-images.githubusercontent.com/24272544/160627913-df402b71-dfb1-4d55-b877-7ebd2a46026e.png)

![multisel_morpheme_derivHist](https://user-images.githubusercontent.com/24272544/160627960-6eae0254-6af6-4aeb-a673-bad6b9998d79.png)

![derivationhistories_list](https://user-images.githubusercontent.com/24272544/160628130-ef9ef851-c979-497d-8bff-78e32b0ac9a1.png)

![derivationhistory_detail_circle](https://user-images.githubusercontent.com/24272544/160628172-4c4bdeed-52f3-4941-a2e1-0d73080cd122.png)

**Comment by susanodd** [comment]
Derivation History is now multiselect.
@Woseseltops I'm going to push this to the same branch since you haven't started reviewing it yet.

**Comment by susanodd** [comment]
This has been pushed to the same branch as semantic fields, now awaiting review and merging.

**Comment by susanodd** [comment]
@Woseseltops merged this into the master branch yesterday.
Once this is live, some additional pages shall be added to the menu under Settings for viewing the new multiple choice fields and their translations for each choice.
Additionally, an Admin (superuser) command will be shown for initializing the Derivation History choices if not yet used by the particular Signbank installation.

**Comment by susanodd** [comment]
Here's the semantic field multiselect shown on glosses in tstMH, with a search on '`Cooking`' (on Global).

![search_semantic_field_cooking](https://user-images.githubusercontent.com/24272544/167407319-90fd9457-198d-427f-ba33-396e24d5238d.png)

![gloss_tstmh_snijden_cooking_fruit](https://user-images.githubusercontent.com/24272544/167407359-06098ebb-e181-4258-bcf9-2edbc6697712.png)

![gloss_tstmh_pompoen_cooking_colour_shape](https://user-images.githubusercontent.com/24272544/167407395-7ac80a5e-175e-4cb8-bb88-597b6b69b3ea.png)

**Comment by susanodd** [comment]
This is live.

**Comment by susanodd** [comment]
For Global, SemanticField and DerivationHistory have model translations for the interface languages (#658)

**Comment by susanodd** [comment]
(Included in #658)
I revised the tests for Revision History so that they use the multiselect SemanticField and DerivationHistory instead of FieldChoice. (Previously they had not been included in the Gloss History, but were merely "skipped over" by using the old FieldChoice model rather than the multiselect model.)
To accomplish this, the update methods for these classes have been updated to save in the GlossRevision. These also show up in the Gloss Revision View in their multiselect version.
